RecoEgamma/EgammaTools/interface/HGCalShowerShapeHelper.h

```diff
@@ -0,0 +1,131 @@
+#include "FWCore/Utilities/interface/StreamID.h"
+#include "Geometry/CaloTopology/interface/HGCalTopology.h"
+#include "Geometry/Records/interface/IdealGeometryRecord.h"
+#include "RecoLocalCalo/HGCalRecAlgos/interface/RecHitTools.h"
+#include "RecoParticleFlow/PFClusterProducer/interface/InitialClusteringStepBase.h"
+
+#include <CLHEP/Vector/LorentzVector.h>
+
+#include <algorithm>
+#include <cstdlib>
+#include <iostream>
+#include <map>
+#include <string>
+#include <type_traits>
+#include <utility>
+#include <vector>
+
+#include <TMatrixD.h>
+#include <TVectorD.h>
+
+#include <Math/Point3D.h>
+#include <Math/Point3Dfwd.h>
+
+class HGCalShowerShapeHelper {
+private:
+  // Good to filter/compute/store this stuff beforehand as they are common to the shower shape variables.
+  // No point in filtering, computing layer-wise centroids, etc. for each variable again and again.
+  // Once intitialized, one can the calculate different variables one after another for a given object.
+  // If a different set of preselections (E, ET, etc.) is required for a given object, then reinitialize using initPerObject(...).
+
+  // In principle should consider the HGCalHSi and HGCalHSc hits (leakage) also.
+  // Can have subdetector dependent thresholds and layer selection.
+  // To be implemented.
+
+  double minHitE_;
+  double minHitET_;
+  double minHitET2_;
+  int minLayer_;
+  int maxLayer_;
+  int nLayer_;
+  DetId::Detector subDet_;
+
+  hgcal::RecHitTools recHitTools_;
+
+  std::unordered_map<uint32_t, const reco::PFRecHit *> pfRecHitPtrMap_;
+  std::vector<std::pair<DetId, float> > hitsAndFracs_;
+  std::vector<double> hitEnergies_;
+  std::vector<double> hitEnergiesWithFracs_;
+
+  ROOT::Math::XYZVector centroid_;
+  std::vector<double> layerEnergies_;
+  std::vector<ROOT::Math::XYZVector> layerCentroids_;
+
+  void setPFRecHitPtrMap(const std::vector<reco::PFRecHit> &recHits);
+
+  void setFilteredHitsAndFractions(const std::vector<std::pair<DetId, float> > &hitsAndFracs);
+
+public:
+  static constexpr double kLDWaferCellSize_ = 0.698;
+  static constexpr double kHDWaferCellSize_ = 0.465;
+
+  void setLayerWiseStuff();
+
+  struct ShowerWidths {
+    double sigma2xx;
+    double sigma2yy;
+    double sigma2zz;
+
+    double sigma2xy;
+    double sigma2yz;
+    double sigma2zx;
+
+    double sigma2uu;
+    double sigma2vv;
+    double sigma2ww;
+
+    ShowerWidths()
+        : sigma2xx(0.0),
+          sigma2yy(0.0),
+          sigma2zz(0.0),
+          sigma2xy(0.0),
+          sigma2yz(0.0),
+          sigma2zx(0.0),
+          sigma2uu(0.0),
+          sigma2vv(0.0),
+          sigma2ww(0.0) {}
+  };
+
+  void initPerEvent(const edm::EventSetup &iSetup, const std::vector<reco::PFRecHit> &recHits);
+
+  void initPerObject(const std::vector<std::pair<DetId, float> > &hitsAndFracs,
+                     double minHitE = 0,
+                     double minHitET = 0,
+                     int minLayer = 1,
+                     int maxLayer = 28,
```

Comment:
can `28` be taken from somewhere else in HGCAL ?

Reply:
@rovere Is the number of HGCal-EE layers stored somewhere?

Reply:
there are methods inside the `RecHitTools` class.

* 28: `rechittools->lastLayerEE()`
* Silicon thickness index: `rechittools->getSiThickIndex(const DetId&)`, see [here](https://github.com/cms-sw/cmssw/blob/206d379168e3af9ee33d63768f3a2996f53e3f7d/DataFormats/ForwardDetId/interface/HGCSiliconDetId.h#L17-L19) and [here, page 260](https://indico.cern.ch/event/977912/contributions/4144796/attachments/2163910/3652620/HGCalGeometry01.pdf)

Reply:
Thanks!